### PR TITLE
Add `target` param to `{Schedule,Sensor}Definition`

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Add new `target` param to `{Schedule,Sensor}Definition` and their associated decorators. A target is an executable entity that can be targeted by a schedule/sensor. This is intended as a replacement for the `job`/`job_name` parameters.

This PR builds off downstack work that consolidated `DirectTarget` and `RepoRelativeTarget` into a single `AutomationTarget` class. The primary addition here is an `AutomationTarget.from_coercible` class method that will accept an `AssetsDefinition` or `CoercibleToAssetSelection` and generate an anonymous asset job from it which is used as the job def for the `AutomationTarget`.

`AssetsDefinitions` that are passed directly as a target, rather than as `Definitions.assets`, are automatically included in the pool of assets for the `Definitions` object. This means that if asset "foo" is embedded as an automation target as some sensor, and an unrelated job is defined by `AssetSelection.assets("foo")`, the job will successfully resolve foo.

## How I Tested These Changes

New unit tests.

### CODE REVIEW

**Comment by smackesey** [comment]
* **#22569** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22569?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py [hunk 1]
```diff
@@ -23,9 +23,10 @@
     SensorDefinition,
     SkipReason,
 )
-from ..target import ExecutableDefinition
+from ..target import CoercibleToAutomationTarget, ExecutableDefinition
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py:26]: I don't think `CoercibleToAutomationTarget` should be part of our public API surface area.

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py [hunk 2]
```diff
@@ -267,46 +268,67 @@ def build_caching_repository_data_from_list(
...
 
+    assets_without_keys = [ad for ad in assets_defs if not ad.keys]
+    asset_graph = AssetGraph.from_assets(
+        [
+            *assets_defs_by_key.values(),
+            *assets_without_keys,
+            *asset_checks_defs,
+            *source_assets_by_key.values(),
+        ]
+        # [*assets_defs, *asset_checks_defs, *source_assets_by_key.values()]
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:319]: rm?

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py [hunk 3]
```diff
@@ -417,7 +439,7 @@ def build_caching_repository_data_from_dict(
     )
 
 
-def _process_and_validate_target(
+def _process_and_validate_target_job(
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:442]: not your fault but this function is brutal

File: python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py:26]: hmm... you think we just inline a long `Union` for the `target` param?

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py (hunk 3 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:442]: Yes the mutating of the passed in dicts is kinda hard to follow. Perhaps worth a refactor in a follow-up.

File: python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py (hunk 1 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py:26]: Yeah just inline the union for now

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py [hunk 4]
```diff
@@ -489,9 +543,33 @@ def _validate_partitions_definition(partitions_def: PartitionsDefinition) -> Non
         )
 
 
-def _get_error_msg_for_target_conflict(targeter, target_type, target_name, dupe_target_type):
+def _get_automation_str(
+    schedule_or_sensor_def: Union[
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:547]: hmmm should we call these `automation_def` by default?

File: python_modules/dagster/dagster/_core/definitions/schedule_definition.py [hunk 5]
```diff
@@ -530,7 +532,7 @@ def with_updated_job(self, new_job: ExecutableDefinition) -> "ScheduleDefinition
         return ScheduleDefinition.dagster_internal_init(
             name=self.name,
             cron_schedule=self._cron_schedule,
-            job_name=self.job_name,
+            job_name=None,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/schedule_definition.py:535]: is this on purpose? If so needs a comment for sure

File: python_modules/dagster/dagster/_core/definitions/target.py [hunk 6]
```diff
@@ -48,14 +59,58 @@ def __new__(
...
+            ),
         )
 
-        return super().__new__(cls, job_resolvable, op_selection=op_selection)
+    @classmethod
+    def from_coercible(
+        cls,
+        coercible: CoercibleToAutomationTarget,
+        automation_name: Optional[str] = None,
+    ) -> Self:
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/target.py:84]: the optional name is worthy of comment

**Review by schrockn — CHANGES_REQUESTED** [review]
A few inline comments worthy of a spin.

File: python_modules/dagster/dagster/_core/definitions/schedule_definition.py (hunk 5 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/schedule_definition.py:535]: yes, I've added a comment. I added logic to make sure you can't simultaneously pass `job_name` and `job` (which should've been there anyway), and in this case the job name will be sourced from the new job in this method

File: python_modules/dagster/dagster/_core/definitions/decorators/schedule_decorator.py [hunk 7]
```diff
@@ -41,6 +52,15 @@ def schedule(
     job: Optional[ExecutableDefinition] = None,
     default_status: DefaultScheduleStatus = DefaultScheduleStatus.STOPPED,
     required_resource_keys: Optional[Set[str]] = None,
+    target: Optional[
+        Union[
+            "CoercibleToAssetSelection",
+            "AssetsDefinition",
+            "JobDefinition",
+            "UnresolvedAssetJobDefinition",
+            "GraphDefinition",
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/schedule_decorator.py:61]: should we still be taking a `GraphSelection` here?

File: python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py [hunk 8]
```diff
@@ -37,6 +46,15 @@ def sensor(
     default_status: DefaultSensorStatus = DefaultSensorStatus.STOPPED,
     asset_selection: Optional[CoercibleToAssetSelection] = None,
     required_resource_keys: Optional[Set[str]] = None,
+    target: Optional[
+        Union[
+            "CoercibleToAssetSelection",
+            "AssetsDefinition",
+            "JobDefinition",
+            "UnresolvedAssetJobDefinition",
+            "GraphDefinition",
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py:55]: same q

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py [hunk 9]
```diff
@@ -417,8 +438,8 @@ def build_caching_repository_data_from_dict(
     )
 
 
-def _process_and_validate_target(
-    schedule_or_sensor_def: Union[
+def _process_and_validate_target_job(
+    automation_def: Union[
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:442]: in our internal code paths at the moment do we not say "instigator" more? I'd rather stay consistent with that for now

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py [hunk 10]
```diff
@@ -457,11 +478,44 @@ def _process_and_validate_target(
...
+                existing_ad = assets_defs_by_key.get(key)
+                if not existing_ad:
+                    assets_defs_by_key[key] = ad
+                elif not existing_ad.computation == ad.computation:
+                    raise DagsterInvalidDefinitionError(
+                        _get_error_msg_for_target_asset_conflict(
+                            _get_automation_str(automation_def), key.to_user_string()
+                        )
+                    )
+            else:  # SourceAsset
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:502]: invariant this. don't rely on a comment

File: python_modules/dagster/dagster/_core/definitions/schedule_definition.py [hunk 11]
```diff
@@ -584,16 +601,36 @@ def __init__(
...
+        elif job:
+            self._target = AutomationTarget.from_coercible(job)
         elif job_name:
             self._target = AutomationTarget(
                 resolvable_to_job=check.str_param(job_name, "job_name"),
-                op_selection=None,
             )
         else:
-            check.failed("Must provide job or job_name")
+            check.failed("Must provide target, job, or job_name")
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/schedule_definition.py:633]: Extracting this into helper that returns a target could be cleaner. Not a necessary change so take it or leave it.

File: python_modules/dagster/dagster/_core/definitions/schedule_definition.py [hunk 12]
```diff
@@ -648,7 +685,7 @@ def _default_run_config_fn(context: ScheduleEvaluationContext) -> CoercibleToRun
                     " to ScheduleDefinition. Must provide only one of the two."
                 )
             elif tags:
-                tags = normalize_tags(tags, allow_reserved_tags=False).tags
+                tags = normalize_tags(tags, allow_reserved_tags=False, warning_stacklevel=5).tags
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/schedule_definition.py:688]: unrelated? Probably positive change but generally like isolating stuff

**Review by schrockn — CHANGES_REQUESTED** [review]
Overall pretty good. Just want to do one more spin quick. Let's explicitly commit to automation or instigator.

File: python_modules/dagster/dagster/_core/definitions/decorators/schedule_decorator.py (hunk 7 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/decorators/schedule_decorator.py:61]: Could go either way but I think so. `GraphDefinition` is still accepted by `job` argument (deprecated for removal in 2.0). Keeps things simpler internally and from a UX perspective to just have `target` be a superset of job rather than also excluding this rarely used case that will be removed later anyway.

File: python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py (hunk 8 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/decorators/sensor_decorator.py:55]: see above response

File: python_modules/dagster/dagster/_core/definitions/schedule_definition.py (hunk 12 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/schedule_definition.py:688]: Not unrelated-- the use of `@experimental_param` (for `target`) on the constructor adds a level to the stack, so adjusting the `warning_stacklevel` keeps the warning message correct.

File: python_modules/dagster/dagster/_core/definitions/schedule_definition.py (hunk 11 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/schedule_definition.py:633]: Because the helper would need to take 4 params I prefer to keep it inline.

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py (hunk 10 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:502]: 👍 fixed

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py (hunk 9 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:442]: older codepaths definitely use "instigator". I was thinking `automation_def` here for consistency with `AutomationTarget` but on reflection I think you're right and we should just use `instigator_def` everywhere.

File: python_modules/dagster/dagster/_core/definitions/decorators/schedule_decorator.py (hunk 7 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/decorators/schedule_decorator.py:61]: I think we should just remove it since it is a greenfield argument.

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py (hunk 9 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py]: oops forgot to make it but just pushed it

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py [hunk 13]
```diff
@@ -490,9 +546,33 @@ def _validate_partitions_definition(partitions_def: PartitionsDefinition) -> Non
...
+        if isinstance(
+            instigator_def,
+            (ScheduleDefinition, UnresolvedPartitionedAssetScheduleDefinition),
+        )
+        else f"sensor '{instigator_def.name}'"
+    )
+
+
+def _get_error_msg_for_target_conflict(
+    automation: str, target_type: str, target_name: str, dupe_target_type: str
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:565]: s/automation/instigator/g

File: python_modules/dagster/dagster/_core/definitions/schedule_definition.py (hunk 11 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/schedule_definition.py:633]: sgtm

**Review by schrockn — APPROVED** [review]
